server/src/instant/scripts/newsletter.clj

```diff
@@ -0,0 +1,156 @@
+  (let [timestamp (java.time.LocalDateTime/now)
+        error-entry {:timestamp timestamp
+                     :email email
+                     :error-type error-type
+                     :message message}]
+    (swap! errors conj error-entry)))
+
+(defn handle-email-error! [e to]
+  (if (inactive-recipient? e)
+    (add-error! to "INACTIVE_RECIPIENT" "This email address has been marked inactive.")
+    (add-error! to "ERROR" (.getMessage e)))
+  nil)
+
+(defn send! [{:keys [to title textBody htmlBody
+                     ;; alias refers to the email template identifer in postmark
+                     ;; In general you shouldn't need to provide any of these
+                     ;; optional params unless you're sending a custom email
+                     alias from reply-to]
+              :or {alias "instant-news"
+                   from "js@hey.instantdb.com"
+                   reply-to "founders@instantdb.com"}}]
+  (let [model {:title title
+               :textBody textBody
+               :htmlBody htmlBody}
+        body {:From from
+              :To to
+              :ReplyTo reply-to
+              :TemplateAlias alias
+              :TemplateModel model
+              :MessageStream "broadcast"
+              :TrackOpens true
+              :TrackLinks "HtmlAndText"
+              :InlineCss true}]
+    (if-not (config/postmark-send-enabled?)
+      (tracer/with-span! {:name "postmark/send-disabled"
+                          :attributes body}
+        (tracer/record-info!
+         {:name "postmark-disabled"
+          :attributes
+          {:msg
+           "Postmark is disabled, add postmark-token to config to enable"}}))
+      (tracer/with-span! {:name "postmark/send"
+                          :attributes body}
+        (try
+          (clj-http/post
+           "https://api.postmarkapp.com/email/withTemplate"
+           {:coerce :always
+            :as :json
+            :headers {"X-Postmark-Server-Token" (config/postmark-token)
+                      "Content-Type" "application/json"}
+            :body (->json body)})
+          (catch Exception e
+            (handle-email-error! e to)))))))
+
+(defn send-newsletter! [{:keys [title slug live?]}]
+  (let [emails (str/split-lines (read-file base-path emails-path))]
+    (if-not live?
+      (doseq [email emails]
+        (log/info "Simulating sending newsletter to" email))
+      (if @once
+        (throw (Exception. "Already sent newsletter! Set once atom to false to send again."))
+        (do
+          (reset! once true)
+          (doseq [email emails]
+            (let [htmlBody (read-file base-path (html-path slug))
+                  textBody (read-file base-path (txt-path slug))]
+              (send! {:to email
+                      :title title
+                      :htmlBody htmlBody
+                      :textBody textBody}))))))))
+
+(comment
+  ;; 1. Set your params
+  (def params {:title "Instant News - Oct 2024" :slug "oct2024"})
```

Comment:
nit: perhaps we don't provide the slug here, and force the user to write it. This way, they would not accidentally eval the forms and end up sending an old version

Reply:
I like it!